File: org.ganttproject.impex.htmlpdf/src/org/ganttproject/impex/htmlpdf/itext/FontSubstitutionPanel.java
/*
GanttProject is an opensource project management tool.
Copyright (C) 2009 Dmitry Barashev

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/
package org.ganttproject.impex.htmlpdf.itext;

import java.awt.Color;
import java.awt.Component;

import javax.swing.Box;
import javax.swing.DefaultCellEditor;
import javax.swing.JComboBox;
import javax.swing.JLabel;
import javax.swing.JScrollPane;
import javax.swing.JTable;
import javax.swing.table.AbstractTableModel;
import javax.swing.table.DefaultTableCellRenderer;
import javax.swing.table.TableCellRenderer;
import javax.swing.table.TableModel;

import org.ganttproject.impex.htmlpdf.itext.FontSubstitutionModel.FontSubstitution;

public class FontSubstitutionPanel {

    private FontSubstitutionModel myModel;
    private JComboBox myFamiliesComboBox;
    private JLabel myMessage;

    public FontSubstitutionPanel(FontSubstitutionModel fontConfigurationModel) {
        myModel = fontConfigurationModel;
    }

    public Component getComponent() {
        TableModel tableModel = new AbstractTableModel() {
            public int getColumnCount() {
                return 2;
            }
            public int getRowCount() {
                return myModel.getSubstitutions().size();
            }
            public Object getValueAt(int rowIndex, int columnIndex) {
                FontSubstitution substitution = getSubstitution(rowIndex);
                switch (columnIndex) {
                case 0:
                    return substitution.myOriginalFamily;
                case 1:
                    return substitution.getSubstitutionFamily();
                default:
                    assert false;
                    throw new IllegalStateException();
                }
            }
            @Override
            public boolean isCellEditable(int rowIndex, int columnIndex) {
                return columnIndex==1;
            }
            @Override
            public void setValueAt(Object value, int rowIndex, int columnIndex) {
                assert columnIndex==1;
                assert value instanceof String;
                getSubstitution(rowIndex).setSubstitutionFamily((String)value);
                updateFontStatusMessage();
            }
            @Override
            public String getColumnName(int column) {
                switch (column) {
                case 0:
                    return "Theme font";
                case 1:
                    return "Substitution";
                default:
                    assert false;
                    throw new IllegalStateException();
                }
            }

        };
        JTable table = new JTable(tableModel);
        table.setRowHeight(30);

        class CellRendererImpl implements TableCellRenderer {
            private DefaultTableCellRenderer myDefaultRenderer = new DefaultTableCellRenderer();
            public Component getTableCellRendererComponent(
                    JTable table, Object value, boolean isSelected, boolean hasFocus,
                    int row, int column) {
                 JLabel result = (JLabel) myDefaultRenderer.getTableCellRendererComponent(
                         table, value, isSelected, hasFocus, row, column);
                 FontSubstitution substitution = getSubstitution(row);
                 if (!substitution.isResolved() || column==0) {
                     result.setFont(result.getFont().deriveFont(16f));
                 }
                 else {
                     result.setFont(substitution.getSubstitutionFont().deriveFont(16f));
                 }
                 if (substitution.isResolved()) {
                     result.setForeground(Color.BLACK);
                 }
                 else {
                     result.setForeground(Color.RED);
                 }
                 return result;
            }
        }

        myFamiliesComboBox = new JComboBox(myModel.getAvailableSubstitutionFamilies().toArray(new String[0]));
        table.getColumnModel().getColumn(0).setCellRenderer(new CellRendererImpl());

        table.getColumnModel().getColumn(1).setCellRenderer(new CellRendererImpl());
        table.getColumnModel().getColumn(1).setCellEditor(new DefaultCellEditor(myFamiliesComboBox));
        table.getTableHeader().setVisible(true);

        myMessage = new JLabel(getMessageText());
        myMessage.setAlignmentX(0);
        Box result = Box.createVerticalBox();
        result.add(myMessage);
        result.add(Box.createVerticalStrut(5));
        result.add(new JScrollPane(table));
        return result;
    }

    private void updateFontStatusMessage() {
        myMessage.setText(getMessageText());
    }

    private String getMessageText() {
        return myModel.hasUnresolvedFonts() ?
                "<html><p><b>Some fonts used in the selected theme have not been found</b></p>"
                + "<p>You may define substitutions in the table</p></html>"
                : "<html>All fonts have been found.</html>";
    }

    private FontSubstitution getSubstitution(int row) {
        return (FontSubstitution) myModel.getSubstitution(row);
    }
}

File: org.ganttproject.impex.htmlpdf/src/main/java/org/ganttproject/impex/htmlpdf/itext/FontSubstitutionPanel.kt
/*
Copyright 2003-2026 Dmitry Barashev, BarD Software s.r.o.

This file is part of GanttProject, an opensource project management tool.

GanttProject is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

GanttProject is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with GanttProject.  If not, see <http://www.gnu.org/licenses/>.
*/

package org.ganttproject.impex.htmlpdf.itext

import biz.ganttproject.lib.fx.GPTableCell
import biz.ganttproject.lib.fx.vbox
import javafx.beans.property.SimpleStringProperty
import javafx.collections.FXCollections
import javafx.event.EventHandler
import javafx.scene.Parent
import javafx.scene.control.*
import javafx.scene.layout.Priority
import javafx.scene.paint.Color
import javafx.scene.text.Font
import javafx.util.Callback
import org.controlsfx.control.tableview2.TableView2
import org.ganttproject.impex.htmlpdf.itext.FontSubstitutionModel.FontSubstitution

/**
 * Represents a UI panel used for managing font substitutions.
 *
 * The panel displays a table containing theme fonts and corresponding substitutions,
 * allowing users to define or modify substitutions for unresolved fonts.
 *
 * @param myModel The data model that holds font substitutions and their resolution state.
 * @param errorMessage A binding property to hold error messages regarding unresolved fonts.
 */
class FontSubstitutionPanel(private val myModel: FontSubstitutionModel, val errorMessage: SimpleStringProperty) {

    fun getComponentFx(): Parent {
        val substitutions = FXCollections.observableArrayList(myModel.substitutions)
        val table = TableView2(substitutions).also { it.styleClass.add("font-substitution-table") }
        table.isEditable = true

        val themeFontColumn = TableColumn<FontSubstitution, String>("Theme font")
        themeFontColumn.cellValueFactory = Callback { features -> SimpleStringProperty(features.value.myOriginalFamily) }
        themeFontColumn.setCellFactory { FontSubstitutionCell(false) }

        val substitutionColumn = TableColumn<FontSubstitution, String>("Substitution")
        substitutionColumn.cellValueFactory = Callback { features -> SimpleStringProperty(features.value.substitutionFamily) }
        substitutionColumn.setCellFactory {
            FontSubstitutionCell(true, myModel.availableSubstitutionFamilies)
        }
        substitutionColumn.onEditCommit = EventHandler { event ->
            val substitution = event.rowValue
            substitution.substitutionFamily = event.newValue
            updateFontStatusMessage()
            table.refresh() // Refresh to update rendering of other columns if necessary
        }

        table.columns.addAll(themeFontColumn, substitutionColumn)
        table.columnResizePolicy = TableView.CONSTRAINED_RESIZE_POLICY

        updateFontStatusMessage()

        return vbox {
          addClasses("font-substitution-panel")
            add(table, alignment = null, growth = Priority.ALWAYS)
        }
    }

    private fun updateFontStatusMessage() {
        if (myModel.hasUnresolvedFonts()) {
          errorMessage.value = "Some fonts used in the selected theme have not been found\nYou may define substitutions in the table"
        } else {
          errorMessage.value = ""
        }
    }
}


private class FontSubstitutionCell(val isSubstitutionColumn: Boolean, val availableFamilies: List<String>? = null) : GPTableCell<FontSubstitution, String>() {
  private var comboBox: ComboBox<String>? = null

  override fun updateItem(item: String?, empty: Boolean) {
    whenNotEmpty(item, empty) {
      val substitution = tableView.items[index]
      if (isEditing && isSubstitutionColumn) {
        graphic = comboBox ?: createDropdown().also {
          it.selectionModel.selectedItemProperty().addListener { _, _, newValue ->
            commitEdit(newValue)
          }
          comboBox = it
        }
        text = null
      } else {
        text = item
        graphic = null

        // Apply font and color
        val awtFont = if (!substitution.isResolved || !isSubstitutionColumn) {
          null
        } else {
          substitution.substitutionFont
        }

        if (awtFont != null) {
          font = Font.font(awtFont.family, 16.0)
        } else {
          font = Font.font(16.0)
        }

        if (substitution.isResolved) {
          textFill = Color.BLACK
        } else {
          textFill = Color.RED
        }
      }
    }
  }
  override fun startEdit() {
    if (!isSubstitutionColumn) return
    super.startEdit()
    graphic = comboBox ?: createDropdown().also { comboBox = it }
    text = null
  }

  override fun cancelEdit() {
    super.cancelEdit()
    text = item
    graphic = null
  }

  private fun createDropdown() = ComboBox(FXCollections.observableArrayList(availableFamilies)).also { comboBox ->
    comboBox.maxWidth = Double.MAX_VALUE
    comboBox.setOnAction {
      commitEdit(comboBox.selectionModel.selectedItem)
    }
    comboBox.selectionModel.select(item)
  }
}

